api/middleware/rateLimiter.go

```diff
@@ -0,0 +1,106 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/api/shared"
+	"github.com/gin-gonic/gin"
+)
+
+type rateLimiter struct {
+	requestsMap    map[string]uint64
+	mutRequestsMap sync.RWMutex
+	limits         map[string]uint64
+	mutLimits      sync.RWMutex
+	countDuration  uint64
+}
+
+// NewRateLimiter returns a new instance of rateLimiter
+func NewRateLimiter(limits map[string]uint64, countDuration uint64) (*rateLimiter, error) {
+	if limits == nil {
+		return nil, ErrNilLimitsMapForEndpoints
+	}
+	return &rateLimiter{
+		requestsMap:   make(map[string]uint64),
+		limits:        limits,
+		countDuration: countDuration,
+	}, nil
+}
+
+// MiddlewareHandlerFunc returns the gin middleware for limiting the number of requests for a given endpoint
+func (rl *rateLimiter) MiddlewareHandlerFunc() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		endpoint := c.FullPath()
+
+		limitForEndpoint, isEndpointLimited := rl.getEndpointLimit(endpoint)
+		if !isEndpointLimited {
+			return
+		}
+
+		clientIP := c.ClientIP()
+		key := fmt.Sprintf("%s_%s", endpoint, clientIP)
+
+		numRequests := rl.loadFromRequestsMap(key)
```

Comment:
might merge the functionality between `loadFromRequestsMap` and `addInRequestsMap` functions as that merge will ensure the atomic operation and will slightly optimize the process.

Reply:
done